messaging/src/test/java/org/axonframework/tracing/IntermediateSpanFactoryTest.java

```diff
@@ -0,0 +1,97 @@
+package org.axonframework.tracing;
+
+import org.axonframework.common.AxonConfigurationException;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.engine.TestDescriptor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Abstract class test for all intermediate factories that result in a {@link Span}. Contains facilities
+ * to setup tests in a good manner and run them.
+ *
+ * @param <BI> The builder class
+ * @param <SI> The span factory class
+ * @since 4.9.0
+ * @author Mitchell Herrijgers
+ */
+public abstract class IntermediateSpanFactoryTest<BI, SI> {
+    private final SpanFactory spanFactory = new TestSpanFactory();
+
+    /**
+     * Creates a new factory builder that also sets the provided span factory.
+     */
+    protected abstract BI createBuilder(SpanFactory spanFactory);
+
+    /**
+     * Executes the builder method and returns the factory based on the builder.
+     */
+    protected abstract SI createFactoryBasedOnBuilder(BI builder);
+
+    @Test
+    void cannotSetNullSpanFactory() {
+        Assertions.assertThrows(AxonConfigurationException.class, () -> createBuilder(null));
+    }
+
+    protected void test(Function<BI, BI> builderAdditions, Function<SI, Span> invocation, SpanDefinition definition) {
+        test(new SpanFactoryTestDefinition(builderAdditions, invocation, definition));
+    }
+
+    protected void test(SpanFactoryTestDefinition definition) {
+        BI builder = createBuilder(spanFactory);
+        builder = definition.getBuilderAdditions().apply(builder);
+        SI spanFactory = createFactoryBasedOnBuilder(builder);
+        Span span = definition.getInvocation().apply(spanFactory);
+        definition.getDefinition().assertSpan((TestSpanFactory.TestSpan) span);
+    }
+
+    public class SpanFactoryTestDefinition {
+        private final Function<BI, BI> builderAdditions;
+        private final Function<SI, Span> invocation;
+        private final SpanDefinition definition;
+
+        public SpanFactoryTestDefinition(Function<BI, BI> builderAdditions, Function<SI, Span> invocation, SpanDefinition definition) {
+            this.builderAdditions = builderAdditions;
+            this.invocation = invocation;
+            this.definition = definition;
+        }
+
+        public Function<BI, BI> getBuilderAdditions() {
+            return builderAdditions;
+        }
+
+        public Function<SI, Span> getInvocation() {
+            return invocation;
+        }
+
+        public SpanDefinition getDefinition() {
+            return definition;
+        }
+    }
+
+    public static class SpanDefinition {
+        private final String name;
```

Comment:
Nit: What about using terminology like "expected" and "actual" on the fields and parameters, to further clarify their relation to one another?

Reply:
Nice changes, thanks!